src/core/reaction_ensemble.cpp

```diff
@@ -0,0 +1,2061 @@
+	}
+	for (int i =0; i<len_reactant_types;i++){
+		bool reactant_type_i_is_known=is_in_list(reactant_types[i],current_reaction_system.type_index,current_reaction_system.nr_different_types);
+		if (reactant_type_i_is_known==false){
+			current_reaction_system.type_index=(int*) realloc(current_reaction_system.type_index, sizeof(int)*(current_reaction_system.nr_different_types+1));
+			current_reaction_system.type_index[current_reaction_system.nr_different_types]=reactant_types[i];
+			current_reaction_system.nr_different_types+=1;
+			status_gc_init_temp=init_type_array(reactant_types[i]); //make types known in espresso
+			status_gc_init=status_gc_init || status_gc_init_temp;
+		}
+	}
+	for (int i =0; i<len_product_types;i++){
+		bool product_type_i_is_known=is_in_list(product_types[i],current_reaction_system.type_index,current_reaction_system.nr_different_types);
+		if (product_type_i_is_known==false){
+			current_reaction_system.type_index=(int*) realloc(current_reaction_system.type_index, sizeof(int)*(current_reaction_system.nr_different_types+1));
+			current_reaction_system.type_index[current_reaction_system.nr_different_types]=product_types[i];
+			current_reaction_system.nr_different_types+=1;
+			status_gc_init_temp=init_type_array(product_types[i]); //make types known in espresso
+			status_gc_init=status_gc_init || status_gc_init_temp;
+		}
+	}
+	
+	//increase current_reaction_system.charges_of_types length
+	current_reaction_system.charges_of_types =(double*) realloc(current_reaction_system.charges_of_types,sizeof(double)*current_reaction_system.nr_different_types);
+	current_reaction_system.charges_of_types[current_reaction_system.nr_different_types-1]=invalid_charge;
+	return status_gc_init;
+}
+
+int find_index_of_type(int type){
+	int index =-100; //initialize to invalid index
+	for(int i=0; i<current_reaction_system.nr_different_types;i++){
+		if(type==current_reaction_system.type_index[i]){
+			index=i;
+			break;
+		}
+	}
+	if(index<0){
+		throw invalid_index();
+		
+	}
+	return index;
+}
+
+
+int replace(int p_id, int desired_type){
+	int err_code_type=set_particle_type(p_id, desired_type);
+	int err_code_q;
+	#ifdef ELECTROSTATICS
+	err_code_q=set_particle_q(p_id, (double) current_reaction_system.charges_of_types[find_index_of_type(desired_type)]);
+	#endif
+	return (err_code_q bitor err_code_type);
+}
+
+
+int hide_particle(int p_id, int previous_type){
+	//remove_charge and put type to a non existing one --> no interactions anymore (not even bonds contribute to energy) it is as if the particle was non existing
+	#ifdef ELECTROSTATICS
+	//set charge
+	set_particle_q(p_id, 0.0);
+	#endif
+	//set type
+	int desired_type=previous_type+100;//+100 in order to assign types that are out of the "usual" range of types
+	int err_code_type=set_particle_type(p_id, desired_type);
+	return err_code_type;
+}
+
+
+//copies last particle to the particle with id p_id and deletes the then last one
+int delete_particle (int p_id) {
```

Comment:
This hinders composablity. Also max_seen_particle does not need to take part in any grand canonical stuff. There is no reason why you would only have polymers and ions in your system. Please come up with a different solution.